<img src="../logo.png" style="float: left;">

# 6. Задачі

## 1. Рахуємо числа (покращено)
Програма приймає від користувача числа. На відміну від попередньої версії, **усі числа вводяться в одному рядку, розділені пробілами**. Програма має рахувати і виводити на екран кількість додатних і від'ємних чисел, суму і середнє значення усіх чисел. Виведення відбувається один раз у кінці запуску. Нуль не вважається ні додатним, ні від'ємним числом.

In [1]:
%run solutions/рахуємо_числа_покращено.py

Введіть усі числа, розділені пробілами: 1 2 3
Додатніх чисел: 3; від'ємних чисел: 0
Сума усіх чисел: 6; середнє значення: 2.0


In [2]:
%run solutions/рахуємо_числа_покращено.py

Введіть усі числа, розділені пробілами: -1 0 1 10 -90
Додатніх чисел: 2; від'ємних чисел: 2
Сума усіх чисел: -80; середнє значення: -16.0


In [3]:
%run solutions/рахуємо_числа_покращено.py

Введіть усі числа, розділені пробілами: 
Додатніх чисел: 0; від'ємних чисел: 0
Сума усіх чисел: 0; середнє значення: не визначено


[Рішення попередньої версії задачі](../у05_loops/solution_notebooks/рахуємо_числа.ipynb) (≈21 рядків коду)   
[Рішення](solution_notebooks/рахуємо_числа_покращено.ipynb) (≈20 рядків коду) 

## 2. Cума цифр числа (покращено, версія 2)
Напишіть програму яка рахує суму цифр числа. Зробіть це простіше використовуючи елементи Пайтон з поточного уроку.

In [6]:
%run solutions/сума_цифр_числа_ще_покращено.py

Введіть число: 939389383784
Фінальна сума цифр: 74


In [7]:
%run solutions/сума_цифр_числа_ще_покращено.py

Введіть число: 1111111111
Фінальна сума цифр: 10


In [8]:
%run solutions/сума_цифр_числа_ще_покращено.py

Введіть число: 0
Фінальна сума цифр: 0


In [9]:
%run solutions/сума_цифр_числа_ще_покращено.py

Введіть число: 
Фінальна сума цифр: 0


[Рішення попередньої версії задачі](../у05_loops/solution_notebooks/сума_цифр_числа_покращено.ipynb) (≈10 рядків коду)   
[Підказка - словами описаний алгоритм](hints_notebooks/сума_цифр_числа_ще_покращено_1.ipynb)  
[Рішення](solution_notebooks/сума_цифр_числа_ще_покращено.ipynb) (≈5 рядків коду) 

## 3. Монети (покращено, версія 2) 
Пам'ятаєте задачу про монети?

Напишіть програму яка приймає від користувача певну суму у копійках і виводить як розбити цю суму на найменшу кількість монет. Для цієї задачі вважатимемо що існують монети у 50, 25, 10, 5 копійок і 1 копійка.

Якщо кількість певної монети нуль, ми не будемо виводити інформацію про неї. Якщо користувач вводить нульову або меншу за нуль суму, необхідно показати повідомлення про помилку.

Зробіть це простіше використовуючи елементи Пайтон з поточного уроку.

In [8]:
%run solutions/монети_ще_покращено.py

Введіть суму: 12341
Монет по 50 копійок: 246
Монет по 25 копійок: 1
Монет по 10 копійок: 1
Монет по 5 копійок: 1
Монет по 1 копійок: 1


[Рішення попередньої версії задачі](../у04_booleans_conditionals/solution_notebooks/монети_покращено.ipynb) (≈22 рядків коду)    
[Підказка](hints_notebooks/монети_ще_покращено_1.ipynb)  
[Рішення](solution_notebooks/монети_ще_покращено.ipynb) (≈10 рядків коду)  

## 4. Прості числа (покращено, версія 2)
Переберіть усі числа від 2 до 20_000, перевірте які з чисел прості, виведіть кожне двохсоте число (якщо виводити кожне то буде забагато чисел). Не використовуйте списки, лише `range()`. Заміряйте скільки часу займає програма. Числа друкуйте в один рядок.

Як можете бачити перше виведене число - 1223. Це означає що перші 199 простих чисел (2, 3, 5, 7 і тд) були пропущені.

In [11]:
%run solutions/прості_числа_повільні.py

1223 2741 4409 6133 7919 9733 11657 13499 15401 17389 19423 
Пошук простих чисел зайняв 1.39 секунд


Тепер давайте оптимізуємо цей пошук використовуючи списки.  

Можна помітити, що коли ми пробігаємося по усім можливим дільникам за допомогою `range()` ми робимо багато зайвої роботи. Наприклад, якщо ми вже встановили що число не ділиться на 2, то немає жодного сенсу перевіряти чи воно ділиться на 4, 6, 8, 10, 12 і тд. Якщо число не ділиться на 11, то немає сенсу перевіряти чи ділиться воно на 22, 33, 44, 55, 66 і тд.  

Тому при перевірці ділиться чи ні треба використовувати лише вже знайдені прості числа. Наприклад, коли ми дійдемо до числа 17, ми будемо перевіряти чи воно ділиться на 2, 3, 5, 7, 11, 13, замість того щоб брати усі числа від 2 до 17.

Перепишіть програму таким чином щоб ми завжди намагалися ділити лише на прості числа і знов поміряйте час виконання.

In [14]:
%run solutions/прості_числа_покращено.py

1223 2741 4409 6133 7919 9733 11657 13499 15401 17389 19423 
Пошук простих чисел зайняв 0.16 секунд


(Насправді тут можна виконати ще одну оптимізацію, але то менш цікаво).

[Рішення попередньої версії задачі](../у05_loops/solution_notebooks/прості_числа.ipynb) (≈16 рядків коду)     
[Підказка - повільне рішення](solution_notebooks/прості_числа_повільні.ipynb) (≈16 рядків коду)    
[Рішення](solution_notebooks/прості_числа_покращено.ipynb) (≈16 рядків коду)  

## 5. Випадкові шестизначні
Згенеруйте випадковим чином і надрукуйте 6 шестизначних чисел у яких не повторюються цифри. Наприклад, 123456 підходить, а 123455 - ні.

In [20]:
%run solutions/випадкові_шестизначні.py

936741 839012 937480 236794 190538 157092 

In [21]:
%run solutions/випадкові_шестизначні.py

106547 376182 892315 965421 274160 427598 

[Підказка #1](hints_notebooks/випадкові_шестизначні_1.ipynb)   
[Підказка #2](hints_notebooks/випадкові_шестизначні_2.ipynb)   
[Рішення](solution_notebooks/випадкові_шестизначні.ipynb) (≈8 рядків коду)  

## 6. Дивні тризначні (покращено, версія 1)
Виведіть усі тризначні числа які містять у собі лише цифри 0, 1, 2, 4. Виводьте по 10 чисел у рядок. Зробіть це простіше використовуючи елементи Пайтон з поточного уроку.

In [22]:
%run solutions/дивні_тризначні_покращено.py

100 101 102 104 110 111 112 114 120 121 
122 124 140 141 142 144 200 201 202 204 
210 211 212 214 220 221 222 224 240 241 
242 244 400 401 402 404 410 411 412 414 
420 421 422 424 440 441 442 444 

[Рішення попередньої версії задачі](../у05_loops/solution_notebooks/дивні_тризначні.ipynb) (≈17 рядків коду)   
[Підказка](hints_notebooks/дивні_тризначні_покращено_1.ipynb)  
[Рішення](solution_notebooks/дивні_тризначні_покращено.ipynb) (≈10 рядків коду) 

## 7. Камінь-ножиці-папір (покращено)
Напишіть програму яка грає з користувачем у камінь-ножиці-папір: програма “загадує” одне з трьох випадковим чином, користувач вводить одне з трьох і далі визначається переможець.

Покращте частину алгоритму, яка відповідає за визначення переможця, за допомогою списків. Приклад роботи програми я не приводжу, тому що він є у [задачах з четвертого уроку](../у04_booleans_conditionals/exercises_booleans_conditionals.ipynb).

[Рішення попередньої версії задачі](../у04_booleans_conditionals/solution_notebooks/камінь_ножиці_папір.ipynb) (≈27 рядків коду)    
[Підказка - фрагмент коду](hints_notebooks/камінь_ножиці_папір_трохи_покращено_1.ipynb)   
[Рішення](solution_notebooks/камінь_ножиці_папір_трохи_покращено.ipynb) (≈24 рядків коду)  

## 8. Паліндром
Для заданого у шаблоні набору рядків чи є кожен з них паліндромом. Рядки-паліндроми це такі рядки які симетричні відносно середини (середньої літери або середніх двох літер якщо довжина рядку парна).

In [14]:
%run solutions/паліндром.py

абвба: паліндром
абввба: паліндром
123421: не паліндром
78a87: паліндром
дамблдор: не паліндром
6666: паліндром
77777: паліндром
FuFuFFuFuFFuFuFFuFuFFuFuFFuFuFFuFuFFuFuFFuFuFFuFuF: паліндром


[Шаблон із рядками](solution_notebooks/паліндром_шаблон.ipynb) (≈7 рядків коду)     
[Підказка](hints_notebooks/паліндром_1.ipynb)   
[Рішення](solution_notebooks/паліндром.ipynb) (≈11 рядків коду)    

## 9. Конвертація балів
Напишіть програму яка зчитує бали студентів, визначає найвищий бал і далі ставить оцінки за наступною схемою:
- Оцінка A якщо бал >= кращий бал - 10
- Оцінка B якщо бал >= кращий бал - 20
- Оцінка C якщо бал >= кращий бал - 30
- Оцінка D якщо бал >= кращий бал - 40
- Інакше F


In [29]:
%run solutions/конвертація_балів.py

Введіть бали, розділені пробілом: 66 77 88 91 93 67 55 21
У студента #0 66 балів, а оцінка - C
У студента #1 77 балів, а оцінка - B
У студента #2 88 балів, а оцінка - A
У студента #3 91 балів, а оцінка - A
У студента #4 93 балів, а оцінка - A
У студента #5 67 балів, а оцінка - C
У студента #6 55 балів, а оцінка - D
У студента #7 21 балів, а оцінка - F


[Рішення #1](solution_notebooks/конвертація_балів.ipynb) (≈20 рядків коду)   
Якщо ваше рішення (дуже) схоже на Рішення #1, спробуйте його зробити красивіше (коротше).  
[Підказка #1 - як зробити красивіше](hints_notebooks/конвертація_балів_1.ipynb)   
[Підказка #2 - як зробити красивіше](hints_notebooks/конвертація_балів_2.ipynb) (≈17 рядків коду)   
[Рішення #2 - красивіше](solution_notebooks/конвертація_балів_2.ipynb) (≈17 рядків коду)    

## 10. Рахуємо кожне число
Напишіть програму яка рахує скільки разів зустрічається кожне число у кожному з рядків із шаблону. Приклади рядків ви можете бачити внизу: усі числа від 0 до 100 (включно) і розділені пробілами. Не використовуйте словники (`dict`).

In [11]:
%run solutions/рахуємо_кожне_число.py

числа: 0 100 100 88 2 2 2 22 31 48 59
лічильники: 0: 1, 2: 3, 22: 1, 31: 1, 48: 1, 59: 1, 88: 1, 100: 2, 

числа: 66 66 66
лічильники: 66: 3, 

числа: 1 2 3 66 66 66 77 77 77 51 51 51 1 2 3 12 3
лічильники: 1: 2, 2: 2, 3: 3, 12: 1, 51: 3, 66: 3, 77: 3, 

числа: 1 1 1 91 91 91 2 2 2 2 2 2 2 2 2 2 2 2 77 47 31
лічильники: 1: 3, 2: 12, 31: 1, 47: 1, 77: 1, 91: 3, 

числа: 1 1 1 91 91 91 2 2 2 2 2 2 2 2 2 2 2 2 77 47 31 18 18 18 18 18 18 18 18
лічильники: 1: 3, 2: 12, 18: 8, 31: 1, 47: 1, 77: 1, 91: 3, 



[Шаблон](solution_notebooks/рахуємо_кожне_число_шаблон.ipynb) (≈10 рядків коду)    
[Підказка](hints_notebooks/рахуємо_кожне_число_1.ipynb)    
[Рішення](solution_notebooks/рахуємо_кожне_число.ipynb) (≈21 рядків коду)    

## 11. Знайти індекс
Для кожного списку чисел із шаблону знайдіть найбільше число і його індекс. Якщо найбільше число зустрічається кілька разів, поверніть найменший індекс.

In [13]:
%run solutions/знайти_індекс.py

індекс=3, значення=14, список=[11, 10, 3, 14, 10, 7]
індекс=5, значення=15, список=[9, 4, 2, 1, 12, 15, 7, 10, 5, 6, 12, 15, 9, 1]
індекс=4, значення=16, список=[14, 15, 8, 3, 16, 16, 4, 10, 5, 10, 14, 11, 6, 10]
індекс=12, значення=16, список=[4, 3, 2, 9, 1, 14, 14, 3, 5, 7, 12, 6, 16, 14]
індекс=11, значення=16, список=[14, 6, 6, 8, 15, 5, 7, 2, 11, 7, 10, 16, 13, 16, 1, 5]
індекс=7, значення=15, список=[2, 3, 14, 8, 9, 3, 6, 15, 10, 8, 15, 8, 11, 9, 1, 6, 10, 6, 15, 8]
індекс=2, значення=16, список=[10, 11, 16, 1, 3, 1, 7, 12, 11, 11, 14]
індекс=3, значення=16, список=[11, 13, 6, 16, 1, 6, 7, 9, 10, 4]
індекс=2, значення=16, список=[6, 13, 16, 7, 16, 3]
індекс=6, значення=16, список=[9, 7, 4, 8, 8, 2, 16, 5, 4, 11, 3, 13, 10, 2]


[Шаблон](solution_notebooks/знайти_індекс_шаблон.ipynb) (≈10 рядків коду)   
[Рішення](solution_notebooks/знайти_індекс.ipynb) (≈13 рядків коду)   

## 12. Перевірка сортування
Напишіть програму яка перевіряє чи є кожен список із шаблону відсортованим у порядку незменшення.

In [17]:
%run solutions/перевірка_сортування.py

Відсортований: True, список: 1 2 3 4 5 666 666
Відсортований: False, список: 1 2 3 2 1
Відсортований: False, список: -1 1 2 -3
Відсортований: True, список: -1234.99 8 9 10 100500 666666


[Шаблон](solution_notebooks/перевірка_сортування_шаблон.ipynb) (≈11 рядків коду)    
[Підказка](hints_notebooks/перевірка_сортування_1.ipynb)    
[Рішення](solution_notebooks/перевірка_сортування.ipynb) (≈18 рядків коду)    

## 13. Злиття відсортованих списків
Напишіть програму яка зливає два відсортовані списки в один відсортований список. Зробіть це найбільш ефективним способом, тобто просто додати два списки і потім відсортувати не допускається.

Пари відсортованих списків задані у шаблоні.

In [1]:
%run solutions/злиття_відсортованих_списків.py

0 1 2 3, 4 5 6 -> [0, 1, 2, 3, 4, 5, 6]
1 1 3 5, 2 4 6 -> [1, 1, 2, 3, 4, 5, 6]
2 4 6, 1 3 5 6 7 -> [1, 2, 3, 4, 5, 6, 6, 7]
2 4 66, 1 3 66 66 -> [1, 2, 3, 4, 66, 66, 66]
6 66 666, 1 1 1 2 -> [1, 1, 1, 2, 6, 66, 666]
3 10 11 18, 1 8 10 11 13 23 -> [1, 3, 8, 10, 10, 11, 11, 13, 18, 23]


[Шаблон](solution_notebooks/злиття_відсортованих_списків_шаблон.ipynb) (≈11 рядків коду)    
[Підказка #1 - про конвертацію](hints_notebooks/злиття_відсортованих_списків_1.ipynb)    
[Підказка #2 - словами описаний алгоритм злиття](hints_notebooks/злиття_відсортованих_списків_2.ipynb)    
[Підказка #3 - ще про алгоритм злиття](hints_notebooks/злиття_відсортованих_списків_3.ipynb)    
[Підказка #4 - фрагмент коду по конвертації](hints_notebooks/злиття_відсортованих_списків_4.ipynb). Ця підказка не пов'зана з підказками #2 і #3, можете дивитись цю підказку якщо у вас виникли проблеми саме з конвертацію.  
[Підказка #5 - фрагмент коду зі злиттям](hints_notebooks/злиття_відсортованих_списків_5.ipynb)    
[Рішення](solution_notebooks/злиття_відсортованих_списків.ipynb) (≈24 рядків коду)    
[Рішення - складніше](solution_notebooks/злиття_відсортованих_списків_складніше.ipynb) (≈27 рядків коду)    

## 14. Колонки
Напишіть програму яка приймає список чисел і друкує колонки, висоти яких дорівнюють цим числам. Для випадку однозначних чисел зробіть щоб колонки виводилися рівно під числами.

In [15]:
%run solutions/колонки.py

Введіть висоти колонок:
1 2 3 1 1 5 5 6 6 7 7 1 1 
                  $ $     
              $ $ $ $     
          $ $ $ $ $ $     
          $ $ $ $ $ $     
    $     $ $ $ $ $ $     
  $ $     $ $ $ $ $ $     
$ $ $ $ $ $ $ $ $ $ $ $ $ 


In [13]:
%run solutions/колонки.py

Введіть висоти колонок:
1 2 3 4 5 6 7 8 9 8 7 6 5 4 3 2 1 1 1 3 3 4 4 7 7 6 6 6
                $                                       
              $ $ $                                     
            $ $ $ $ $                         $ $       
          $ $ $ $ $ $ $                       $ $ $ $ $ 
        $ $ $ $ $ $ $ $ $                     $ $ $ $ $ 
      $ $ $ $ $ $ $ $ $ $ $               $ $ $ $ $ $ $ 
    $ $ $ $ $ $ $ $ $ $ $ $ $         $ $ $ $ $ $ $ $ $ 
  $ $ $ $ $ $ $ $ $ $ $ $ $ $ $       $ $ $ $ $ $ $ $ $ 
$ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ $ 


Для двозначних чисел програма теж має працювати, але необов'язково так красиво.

In [14]:
%run solutions/колонки.py

Введіть висоти колонок:
11 12 13 13 13 6 6 6 1 2 3 19 19
                      $ $ 
                      $ $ 
                      $ $ 
                      $ $ 
                      $ $ 
                      $ $ 
    $ $ $             $ $ 
  $ $ $ $             $ $ 
$ $ $ $ $             $ $ 
$ $ $ $ $             $ $ 
$ $ $ $ $             $ $ 
$ $ $ $ $             $ $ 
$ $ $ $ $             $ $ 
$ $ $ $ $ $ $ $       $ $ 
$ $ $ $ $ $ $ $       $ $ 
$ $ $ $ $ $ $ $       $ $ 
$ $ $ $ $ $ $ $     $ $ $ 
$ $ $ $ $ $ $ $   $ $ $ $ 
$ $ $ $ $ $ $ $ $ $ $ $ $ 


[Підказка #1](hints_notebooks/колонки_1.ipynb)    
[Підказка #2](hints_notebooks/колонки_2.ipynb)    
[Рішення](solution_notebooks/колонки.ipynb) (≈16 рядків коду)   

## 15. Хрестики-нулики: покрокове заповнення
Напишіть програму, яка спрощено емулює процес гри у хрестики-нулики. У цій спрощеній емуляції комп'ютер буде заповнювати поле для гри почергово хрестиками і нуликами ніби у гру грають два ком'ютерні боти випадковим чином. Ми не будемо визначати переможця і завчасно зупиняти гру якщо один гравець вже переміг. Виводьте на екран усі дев'ять ходів суперників один за одним.

Така спрощена імплементація то перший крок для того щоб зробити повну гру у хрестики-нулики де людина може грати проти комп'ютера. Строго кажучи знань циклів і списків вже достатньо щоб зробити повну імплементацію, але зробити це без функцій це значно складніше, тому я не раджу, хіба що вам усі вправи даються дуже легко.

**Зверніть увагу:** задача стоїть про заповнення саме покроково, хід за ходом. Просто згенерувати випадковим чином фінальний розподіл хрестиків і нуликів коли на ігровому полі коли немає жодної вільної клітинки значно легше, і це **не** є поточною задачею.

In [4]:
%run solutions/хрестики_нулики_заповнення.py

_ _ _ 
_ X _ 
_ _ _ 

_ _ 0 
_ X _ 
_ _ _ 

_ _ 0 
_ X _ 
X _ _ 

0 _ 0 
_ X _ 
X _ _ 

0 _ 0 
_ X _ 
X _ X 

0 _ 0 
_ X _ 
X 0 X 

0 _ 0 
X X _ 
X 0 X 

0 0 0 
X X _ 
X 0 X 

0 0 0 
X X X 
X 0 X 



[Підказка #1 - абстрактна](hints_notebooks/хрестики_нулики_заповнення_1.ipynb)  
[Підказка #2 - словами описаний алгоритм](hints_notebooks/хрестики_нулики_заповнення_2.ipynb) (≈21 рядків коду)  
[Підказка #3 - фрагмент коду](hints_notebooks/хрестики_нулики_заповнення_3.ipynb) (≈12 рядків коду)  
[Підказка #4 - дуже детальний фрагмент коду](hints_notebooks/хрестики_нулики_заповнення_4.ipynb)  
[Рішення](solution_notebooks/хрестики_нулики_заповнення.ipynb) (≈20 рядків коду)    
[Бонусне запитання (після того як розв'яжете задачу)](hints_notebooks/хрестики_нулики_заповнення_5.ipynb)  

Якщо ваше рішення дуже схоже на моє, то тепер я пропоную вирішити задачу трохи красивіше (**тільки після того як вирішите задачу**).
[Пояснення про красивіше](hints_notebooks/хрестики_нулики_заповнення_6.ipynb)  
[Підказка #5 - досить абстрактна](hints_notebooks/хрестики_нулики_заповнення_7.ipynb)  
[Підказка #6 - менш абстрактна](hints_notebooks/хрестики_нулики_заповнення_8.ipynb)  
[Підказка #7 - досить конкретна](hints_notebooks/хрестики_нулики_заповнення_9.ipynb)  
[Підказка #8 - словами описаний алгоритм](hints_notebooks/хрестики_нулики_заповнення_10.ipynb)  
[Підказка #9 - детальний фрагмент коду](hints_notebooks/хрестики_нулики_заповнення_11.ipynb)  
[Рішення #2 - більш прямолінійне](solution_notebooks/хрестики_нулики_заповнення_2.ipynb) (≈21 рядків коду)     

## 16. Хрестики-нулики: фінальне заповнення
По мотивам попередньої вправи "Хрестики-нулики: покрокове заповнення".

Напишіть програму, яка виводить фінальний стан гри двох ботів у хрестики-нулики. Не перевіряйте хто виграв, завжди заповнюйте ігрове поле хрестиками-нуликами повністю. Зробіть програму якомога коротшою.

In [6]:
%run solutions/хрестики_нулики_заповнення_фінальне.py

X 0 0 
X 0 X 
X X 0 


[Підказка #1](hints_notebooks/хрестики_нулики_заповнення_фінальне_1.ipynb)  
[Підказка #2](hints_notebooks/хрестики_нулики_заповнення_фінальне_2.ipynb) (≈5 рядків коду)  
[Рішення #1](solution_notebooks/хрестики_нулики_заповнення_фінальне.ipynb) (≈7 рядків коду)      
[Рішення #2 - ще коротше](solution_notebooks/хрестики_нулики_заповнення_фінальне_2.ipynb) (≈5 рядків коду)    

## 17. Дошка Гальтона
<img src="galton_boards.png"  style="right;">
Дошка Гальтона це прилад для статистичних експериментів, названий на честь вченого Френсіса Гальтона. Вона складається з вертикальної дошки з рівномірно вбитими цвяхами у трикутній формі.

Рекомендую переглянути короткі відео на ютубі щоб краще візуалізувати задачу: [відео 48 секунд](https://www.youtube.com/watch?v=Vo9Esp1yaC8&ab_channel=NamirhaAhriman), [більше драматичне відео <2 хв](https://www.youtube.com/watch?v=9xUBhhM4vbM&ab_channel=IndexFundAdvisors%2CInc), [відео-моделювання 24 секунди](https://www.youtube.com/watch?v=FJE-3aHLIZc&ab_channel=GeneralPhysics-ProblemsandSolutions).

Кульки скидаються з одного отвору що знаходиться зверху і посередені. Кожного разу як кулька вдаряється о цвях, вона має 50% шанс впасти вліво чи вправо. Купки кульок накопичуються внизу дошки у спеціальних слотах.

Напишіть програму, яка моделює результат роботи дошки Гальтона для варіанту з 13 слотами і 266 кульками. Гістограму з кульками можете виводити горизонтально.

Опціонально: [Дошка Гальтона на українській Wikipedia](https://uk.wikipedia.org/wiki/%D0%94%D0%BE%D1%88%D0%BA%D0%B0_%D0%93%D0%B0%D0%BB%D1%8C%D1%82%D0%BE%D0%BD%D0%B0)



In [7]:
%run solutions/дошка_гальтона.py

|*
|
|******
|********************
|*****************************
|**************************************************
|******************************************************
|*******************************************************
|*****************************
|****************
|*****
|*
|


In [12]:
%run solutions/дошка_гальтона.py

|
|**
|********
|********************
|**********************************
|*************************************************
|*******************************************************
|************************************************
|****************************
|***************
|******
|
|*


Пишіть код таким чином, щоб його можна було легко зконфігурувати для роботи з іншими параметрами, наприклад, для 25 слотів і 666 кульок.

In [25]:
%run solutions/дошка_гальтона_інші_параметри.py

|
|
|
|
|*
|*
|******
|****************
|************************************
|***********************************************************
|***********************************************************************
|***********************************************************************************************************
|*****************************************************************************************
|*******************************************************************************************
|***************************************************************************
|******************************************************************
|**************************
|***************
|******
|*
|
|
|
|
|


[Підказка #1](hints_notebooks/дошка_гальтона_1.ipynb) - загальна підказка   
[Підказка #2](hints_notebooks/дошка_гальтона_2.ipynb) - детальніша підказка (без коду)  
[Рішення](solution_notebooks/дошка_гальтона.ipynb) (≈12 рядків коду)    

## 18. Діагоналі
Напишіть програму яка малює усі можливі діагоналі на шаховій дошці. Це хороша підготовка до наступної задачі про Вісім Ферзів.

In [3]:
%run solutions/діагоналі.py


| | | | | | | |D|
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |

| | | | | | |D| |
| | | | | | | |D|
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |

| | | | | |D| | |
| | | | | | |D| |
| | | | | | | |D|
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |

| | | | |D| | | |
| | | | | |D| | |
| | | | | | |D| |
| | | | | | | |D|
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |

| | | |D| | | | |
| | | | |D| | | |
| | | | | |D| | |
| | | | | | |D| |
| | | | | | | |D|
| | | | | | | | |
| | | | | | | | |
| | | | | | | | |

| | |D| | | | | |
| | | |D| | | | |
| | | | |D| | | |
| | | | | |D| | |
| | | | | | |D| |
| | | | | | | |D|
| | | | | | | | |
| | | | | | | | |

| |D| | | | | | |
| | |D| | | | | |
| | | |D| | | | |
| | | | |D| | | |
| | | | | |D| | |
| | | | | | |D| |
| | | | | | | |D|
| |

[Підказка #1](hints_notebooks/діагоналі_1.ipynb) - загальна підказка  
[Підказка #2](hints_notebooks/діагоналі_2.ipynb) - формула діагоналі   
[Підказка #3](hints_notebooks/діагоналі_3.ipynb) - формули діагоналей  
[Підказка #4](hints_notebooks/діагоналі_4.ipynb) - високорівний опис алгоритму текстом  
[Підказка #5](hints_notebooks/діагоналі_5.ipynb) - шаблон програми (Python код)  
[Рішення з дублюванням](solution_notebooks/діагоналі_з_дублюванням.ipynb) (≈34 рядків коду)        . Спробуйте самостійно позбутися дублікації у цьому рішенні.  
[Рішення](solution_notebooks/діагоналі.ipynb) (≈28 рядків коду)   
[Підказка #6](hints_notebooks/діагоналі_6.ipynb) - спробуйте скоротити ще більше  
[Рішення (коротше, але менш красиво)](solution_notebooks/діагоналі_коротше.ipynb) (≈20 рядків коду)    

## 19. Вісім ферзів

<img src="queen_stike.png" style="float: right;">

Задача про вісім ферзів полягає в такому розміщенні восьми ферзів на шахівниці, що жоден з них не ставить під удар один одного. Тобто, вони не повинні стояти на одній вертикалі, горизонталі чи діагоналі. Напишіть програму, яка виводить будь-який один із розв'язків такої задачі.

Знайти усі можливі рішення у такій задачі досить складно, вивести один із варіантів - простіша задача.

[Підказка #1](hints_notebooks/вісім_ферзів_1.ipynb) - невелика підказка  
[Підказка #2](hints_notebooks/вісім_ферзів_2.ipynb) - конкретніша підказка  
[Підказка #3](hints_notebooks/вісім_ферзів_3.ipynb) - підказка з фрагментом робочого коду  
[Підказка #4](hints_notebooks/вісім_ферзів_4.ipynb) - часткове рішення з прогалинами, не запускається  
[Підказка #5](hints_notebooks/вісім_ферзів_5.ipynb) - майже готове часткове рішення, що запускається  
[Рішення](solution_notebooks/вісім_ферзів.ipynb) (≈40 рядків коду)  
[Рішення - без дублікації](solution_notebooks/вісім_ферзів_без_дублікації.ipynb) (≈40 рядків коду)  

In [1]:
%run solutions/вісім_ферзів.py

Пошук зайняв 896 спроб
| | | |Q| | | | |
| |Q| | | | | | |
| | | | | | |Q| |
| | |Q| | | | | |
| | | | | |Q| | |
| | | | | | | |Q|
| | | | |Q| | | |
|Q| | | | | | | |
